editor/src/clj/editor/debugging/mobdebug.clj

```diff
@@ -109,31 +201,114 @@
+        b-type (classify-type b)]
+    (case a-type
+      :number (case b-type
+                :number (compare a b)
+                (:string :comparable :other) -1)
+      :string (case b-type
+                :number 1
+                :string (compare a b)
+                (:comparable :other) -1)
+      :comparable (case b-type
+                    (:number :string) 1
+                    :comparable (compare a b)
+                    :other -1)
+      :other (case b-type
+               (:number :string :comparable) 1
+               :other (compare (System/identityHashCode a)
+                               (System/identityHashCode b))))))
+
+(defn- read-lua-table [{:keys [content string]}]
+  (with-meta (into (sorted-map-by compare-keys)
+                   (sequence->map-entries content))
+             {:string string}))
+
+(declare ->LuaStructure)
+
+(defn- decode-serialized-data
+  [debug-session edn-string]
+  (try
+    (edn/read-string
+      {:readers {'lua/table read-lua-table
+                 'lua/ref ->LuaRef
+                 'lua/structure (fn [{:keys [value refs]}]
+                                  (->LuaStructure debug-session value refs))}
+       :default ->LuaBlackBox}
+      edn-string)
+    (catch Exception e
+      (throw (ex-info "Error decoding serialized data" {:input edn-string} e)))))
+
+;; A representation of lua table that supports circular references. Acts like a
+;; read-only map (no assoc/dissoc) that creates/loads sub-structures on demand.
+
+(defn- maybe-ref->structure [^DebugSession debug-session x refs]
+  (if (instance? LuaRef x)
+    (if (contains? refs x)
+      (->LuaStructure debug-session x refs)
+      ;; We have a reference that was not serialized, load it:
+      (with-session debug-session
+        (when (= :suspended (-state debug-session))
+          (let [out (.-out debug-session)
+                in (.-in debug-session)]
+            (send-command! out (format "REF %s --{maxlevel=4}" (:address x)))
+            (let [[status rest] (read-status in)]
+              (case status
+                "200"
+                (when-let [[edn-string] (re-match #"^OK\s+" rest)]
+                  (let [loaded-refs (::refs (decode-serialized-data debug-session edn-string))]
+                    (->LuaStructure debug-session x (into refs loaded-refs))))
+
+                "401"
+                (when-let [[size] (re-match #"^Error in Execution\s+(\d+)$" rest)]
+                  (let [message (read-data in (Integer/parseInt size))]
+                    (log/warn :message "Failed to serialize the ref"
+                              :ref (:address x)
+                              :error-message message)
+                    nil))
+
+                "400"
+                (do (log/warn :message "Couldn't load lua table for a ref: not registered on a server"
+                              :ref (:address x)
+                              :error-message rest)
+                    nil)))))))
+    x))
```

Comment:
New addition. When we convert `LuaRef` (internal data structure) to `LuaStructure` (public data structure), its contents might be missing because we don't serialize the whole graph. In that case, we load it from the debugged game.